// devextreme/cjs/__internal/grids/grid_core/error_handling/m_error_handling.js
/**
 * DevExtreme (cjs/__internal/grids/grid_core/error_handling/m_error_handling.js)
 * Version: 23.1.13
 * Build date: Mon Oct 28 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.errorHandlingModule = exports.ErrorHandlingController = void 0;
var _renderer = _interopRequireDefault(require("../../../../core/renderer"));
var _iterator = require("../../../../core/utils/iterator");
var _click = require("../../../../events/click");
var _events_engine = _interopRequireDefault(require("../../../../events/core/events_engine"));
var _m_modules = _interopRequireDefault(require("../m_modules"));

function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    }
}

function _inheritsLoose(subClass, superClass) {
    subClass.prototype = Object.create(superClass.prototype);
    subClass.prototype.constructor = subClass;
    _setPrototypeOf(subClass, superClass)
}

function _setPrototypeOf(o, p) {
    _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function(o, p) {
        o.__proto__ = p;
        return o
    };
    return _setPrototypeOf(o, p)
}
var ERROR_ROW_CLASS = "dx-error-row";
var ERROR_MESSAGE_CLASS = "dx-error-message";
var ERROR_CLOSEBUTTON_CLASS = "dx-closebutton";
var ACTION_CLASS = "action";
var ErrorHandlingController = function(_modules$ViewControll) {
    _inheritsLoose(ErrorHandlingController, _modules$ViewControll);

    function ErrorHandlingController() {
        return _modules$ViewControll.apply(this, arguments) || this
    }
    var _proto = ErrorHandlingController.prototype;
    _proto.init = function() {
        this._resizingController = this.getController("resizing");
        this._columnsController = this.getController("columns");
        this._columnHeadersView = this.getView("columnHeadersView");
        this._rowsView = this.getView("rowsView")
    };
    _proto._createErrorRow = function(error, $tableElements) {
        var _this = this;
        var $errorRow;
        var $closeButton;
        var $errorMessage = this._renderErrorMessage(error);
        if ($tableElements) {
            $errorRow = (0, _renderer.default)("<tr>").addClass(ERROR_ROW_CLASS);
            $closeButton = (0, _renderer.default)("<div>").addClass(ERROR_CLOSEBUTTON_CLASS).addClass(this.addWidgetPrefix(ACTION_CLASS));
            _events_engine.default.on($closeButton, _click.name, this.createAction((function(args) {
                var _a, _b;
                var e = args.event;
                var $errorRow;
                var errorRowIndex = (0, _renderer.default)(e.currentTarget).closest(".".concat(ERROR_ROW_CLASS)).index();
                e.stopPropagation();
                (0, _iterator.each)($tableElements, (function(_, tableElement) {
                    $errorRow = (0, _renderer.default)(tableElement).children("tbody").children("tr").eq(errorRowIndex);
                    _this.removeErrorRow($errorRow)
                }));
                null === (_b = null === (_a = _this._resizingController) || void 0 === _a ? void 0 : _a.fireContentReadyAction) || void 0 === _b ? void 0 : _b.call(_a)
            })));
            (0, _renderer.default)("<td>").attr({
                colSpan: this.getController("columns").getVisibleColumns().length,
                role: "presentation"
            }).prepend($closeButton).append($errorMessage).appendTo($errorRow);
            return $errorRow
        }
        return $errorMessage
    };
    _proto._renderErrorMessage = function(error) {
        var message = error.url ? error.message.replace(error.url, "") : error.message || error;
        var $message = (0, _renderer.default)("<div>").addClass(ERROR_MESSAGE_CLASS).text(message);
        if (error.url) {
            (0, _renderer.default)("<a>").attr("href", error.url).text(error.url).appendTo($message)
        }
        return $message
    };
    _proto.renderErrorRow = function(error, rowIndex, $popupContent) {
        var _a, _b;
        var that = this;
        var $errorMessageElement;
        var $firstErrorRow;
        if ($popupContent) {
            $popupContent.find(".".concat(ERROR_MESSAGE_CLASS)).remove();
            $errorMessageElement = that._createErrorRow(error);
            $popupContent.prepend($errorMessageElement);
            return $errorMessageElement
        }
        var viewElement = rowIndex >= 0 || !that._columnHeadersView.isVisible() ? that._rowsView : that._columnHeadersView;
        var $tableElements = viewElement.getTableElements();
        (0, _iterator.each)($tableElements, (function(_, tableElement) {
            $errorMessageElement = that._createErrorRow(error, $tableElements);
            $firstErrorRow = $firstErrorRow || $errorMessageElement;
            if (rowIndex >= 0) {
                var $row = viewElement._getRowElements((0, _renderer.default)(tableElement)).eq(rowIndex);
                that.removeErrorRow($row.next());
                $errorMessageElement.insertAfter($row)
            } else {
                var $tbody = (0, _renderer.default)(tableElement).children("tbody");
                var rowElements = $tbody.children("tr");
                if (that._columnHeadersView.isVisible()) {
                    that.removeErrorRow(rowElements.last());
                    (0, _renderer.default)(tableElement).append($errorMessageElement)
                } else {
                    that.removeErrorRow(rowElements.first());
                    $tbody.first().prepend($errorMessageElement)
                }
            }
        }));
        null === (_b = null === (_a = this._resizingController) || void 0 === _a ? void 0 : _a.fireContentReadyAction) || void 0 === _b ? void 0 : _b.call(_a);
        return $firstErrorRow
    };
    _proto.removeErrorRow = function($row) {
        if (!$row) {
            var $columnHeaders = this._columnHeadersView && this._columnHeadersView.element();
            $row = $columnHeaders && $columnHeaders.find(".".concat(ERROR_ROW_CLASS));
            if (!$row || !$row.length) {
                var $rowsViewElement = this._rowsView.element();
                $row = $rowsViewElement && $rowsViewElement.find(".".concat(ERROR_ROW_CLASS))
            }
        }
        $row && $row.hasClass(ERROR_ROW_CLASS) && $row.remove()
    };
    _proto.optionChanged = function(args) {
        switch (args.name) {
            case "errorRowEnabled":
                args.handled = true;
                break;
            default:
                _modules$ViewControll.prototype.optionChanged.call(this, args)
        }
    };
    return ErrorHandlingController
}(_m_modules.default.ViewController);
exports.ErrorHandlingController = ErrorHandlingController;
var data = function(Base) {
    return function(_Base) {
        _inheritsLoose(ErrorHandlingDataControllerExtends, _Base);

        function ErrorHandlingDataControllerExtends() {
            return _Base.apply(this, arguments) || this
        }
        var _proto2 = ErrorHandlingDataControllerExtends.prototype;
        _proto2.init = function() {
            var _this2 = this;
            _Base.prototype.init.call(this);
            this.dataErrorOccurred.add((function(error, $popupContent) {
                if (_this2.option("errorRowEnabled")) {
                    _this2._errorHandlingController.renderErrorRow(error, void 0, $popupContent)
                }
            }));
            this.changed.add((function(e) {
                var _a, _b;
                if (e && "loadError" === e.changeType) {
                    return
                }
                if (_this2._editingController && !_this2._editingController.hasChanges()) {
                    null === (_b = null === (_a = _this2._errorHandlingController) || void 0 === _a ? void 0 : _a.removeErrorRow) || void 0 === _b ? void 0 : _b.call(_a)
                }
            }))
        };
        return ErrorHandlingDataControllerExtends
    }(Base)
};
var errorHandlingModule = {
    defaultOptions: function() {
        return {
            errorRowEnabled: true
        }
    },
    controllers: {
        errorHandling: ErrorHandlingController
    },
    extenders: {
        controllers: {
            data: data
        }
    }
};
exports.errorHandlingModule = errorHandlingModule;
