// devextreme-angular/esm2022/ui/form/nested/tab-dxi.mjs
/*!
 * devextreme-angular
 * Version: 25.1.3
 * Build date: Wed Jun 25 2025
 *
 * Copyright (c) 2012 - 2025 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
/* tslint:disable:max-line-length */
import { Component, NgModule, Host, ElementRef, Renderer2, Inject, SkipSelf, Input, ContentChildren, forwardRef, QueryList } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import { DxIntegrationModule, NestedOptionHost, extractTemplate, DxTemplateHost } from 'devextreme-angular/core';
import { CollectionNestedOption } from 'devextreme-angular/core';
import { DxiFormButtonItemComponent } from './button-item-dxi';
import { DxiFormEmptyItemComponent } from './empty-item-dxi';
import { DxiFormGroupItemComponent } from './group-item-dxi';
import { DxiFormItemComponent } from './item-dxi';
import { DxiFormSimpleItemComponent } from './simple-item-dxi';
import { DxiFormTabbedItemComponent } from './tabbed-item-dxi';
import * as i0 from "@angular/core";
import * as i1 from "devextreme-angular/core";
export class DxiFormTabComponent extends CollectionNestedOption {
    renderer;
    document;
    element;
    get alignItemLabels() {
        return this._getOption('alignItemLabels');
    }
    set alignItemLabels(value) {
        this._setOption('alignItemLabels', value);
    }
    get badge() {
        return this._getOption('badge');
    }
    set badge(value) {
        this._setOption('badge', value);
    }
    get colCount() {
        return this._getOption('colCount');
    }
    set colCount(value) {
        this._setOption('colCount', value);
    }
    get colCountByScreen() {
        return this._getOption('colCountByScreen');
    }
    set colCountByScreen(value) {
        this._setOption('colCountByScreen', value);
    }
    get disabled() {
        return this._getOption('disabled');
    }
    set disabled(value) {
        this._setOption('disabled', value);
    }
    get icon() {
        return this._getOption('icon');
    }
    set icon(value) {
        this._setOption('icon', value);
    }
    get items() {
        return this._getOption('items');
    }
    set items(value) {
        this._setOption('items', value);
    }
    get tabTemplate() {
        return this._getOption('tabTemplate');
    }
    set tabTemplate(value) {
        this._setOption('tabTemplate', value);
    }
    get template() {
        return this._getOption('template');
    }
    set template(value) {
        this._setOption('template', value);
    }
    get title() {
        return this._getOption('title');
    }
    set title(value) {
        this._setOption('title', value);
    }
    get _optionPath() {
        return 'tabs';
    }
    buttonItemsChildren;
    emptyItemsChildren;
    groupItemsChildren;
    itemsChildren;
    simpleItemsChildren;
    tabbedItemsChildren;
    setItems() {
        const q = new QueryList();
        q.reset([
            ...this.buttonItemsChildren.toArray(),
            ...this.emptyItemsChildren.toArray(),
            ...this.groupItemsChildren.toArray(),
            ...this.itemsChildren.toArray(),
            ...this.simpleItemsChildren.toArray(),
            ...this.tabbedItemsChildren.toArray(),
        ]);
        this.setChildren('items', q);
    }
    constructor(parentOptionHost, optionHost, renderer, document, templateHost, element) {
        super();
        this.renderer = renderer;
        this.document = document;
        this.element = element;
        parentOptionHost.setNestedOption(this);
        optionHost.setHost(this, this._fullOptionPath.bind(this));
        templateHost.setHost(this);
    }
    setTemplate(template) {
        this.template = template;
    }
    ngAfterViewInit() {
        extractTemplate(this, this.element, this.renderer, this.document);
    }
    ngOnDestroy() {
        this._deleteRemovedOptions(this._fullOptionPath());
    }
    ngAfterContentInit() {
        this.setItems();
        this.buttonItemsChildren.changes.subscribe(() => { this.setItems(); });
        this.emptyItemsChildren.changes.subscribe(() => { this.setItems(); });
        this.groupItemsChildren.changes.subscribe(() => { this.setItems(); });
        this.itemsChildren.changes.subscribe(() => { this.setItems(); });
        this.simpleItemsChildren.changes.subscribe(() => { this.setItems(); });
        this.tabbedItemsChildren.changes.subscribe(() => { this.setItems(); });
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.12", ngImport: i0, type: DxiFormTabComponent, deps: [{ token: i1.NestedOptionHost, host: true, skipSelf: true }, { token: i1.NestedOptionHost, host: true }, { token: i0.Renderer2 }, { token: DOCUMENT }, { token: i1.DxTemplateHost, host: true }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Component });
    /** @nocollapse */ static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.12", type: DxiFormTabComponent, isStandalone: true, selector: "dxi-form-tab", inputs: { alignItemLabels: "alignItemLabels", badge: "badge", colCount: "colCount", colCountByScreen: "colCountByScreen", disabled: "disabled", icon: "icon", items: "items", tabTemplate: "tabTemplate", template: "template", title: "title" }, providers: [NestedOptionHost, DxTemplateHost], queries: [{ propertyName: "buttonItemsChildren", predicate: i0.forwardRef(() => DxiFormButtonItemComponent) }, { propertyName: "emptyItemsChildren", predicate: i0.forwardRef(() => DxiFormEmptyItemComponent) }, { propertyName: "groupItemsChildren", predicate: i0.forwardRef(() => DxiFormGroupItemComponent) }, { propertyName: "itemsChildren", predicate: i0.forwardRef(() => DxiFormItemComponent) }, { propertyName: "simpleItemsChildren", predicate: i0.forwardRef(() => DxiFormSimpleItemComponent) }, { propertyName: "tabbedItemsChildren", predicate: i0.forwardRef(() => DxiFormTabbedItemComponent) }], usesInheritance: true, ngImport: i0, template: '<ng-content></ng-content>', isInline: true, styles: [":host{display:block}\n"], dependencies: [{ kind: "ngmodule", type: DxIntegrationModule }] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.12", ngImport: i0, type: DxiFormTabComponent, decorators: [{
            type: Component,
            args: [{ selector: 'dxi-form-tab', standalone: true, template: '<ng-content></ng-content>', imports: [DxIntegrationModule], providers: [NestedOptionHost, DxTemplateHost], styles: [":host{display:block}\n"] }]
        }], ctorParameters: () => [{ type: i1.NestedOptionHost, decorators: [{
                    type: SkipSelf
                }, {
                    type: Host
                }] }, { type: i1.NestedOptionHost, decorators: [{
                    type: Host
                }] }, { type: i0.Renderer2 }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }, { type: i1.DxTemplateHost, decorators: [{
                    type: Host
                }] }, { type: i0.ElementRef }], propDecorators: { alignItemLabels: [{
                type: Input
            }], badge: [{
                type: Input
            }], colCount: [{
                type: Input
            }], colCountByScreen: [{
                type: Input
            }], disabled: [{
                type: Input
            }], icon: [{
                type: Input
            }], items: [{
                type: Input
            }], tabTemplate: [{
                type: Input
            }], template: [{
                type: Input
            }], title: [{
                type: Input
            }], buttonItemsChildren: [{
                type: ContentChildren,
                args: [forwardRef(() => DxiFormButtonItemComponent)]
            }], emptyItemsChildren: [{
                type: ContentChildren,
                args: [forwardRef(() => DxiFormEmptyItemComponent)]
            }], groupItemsChildren: [{
                type: ContentChildren,
                args: [forwardRef(() => DxiFormGroupItemComponent)]
            }], itemsChildren: [{
                type: ContentChildren,
                args: [forwardRef(() => DxiFormItemComponent)]
            }], simpleItemsChildren: [{
                type: ContentChildren,
                args: [forwardRef(() => DxiFormSimpleItemComponent)]
            }], tabbedItemsChildren: [{
                type: ContentChildren,
                args: [forwardRef(() => DxiFormTabbedItemComponent)]
            }] } });
export class DxiFormTabModule {
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.12", ngImport: i0, type: DxiFormTabModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    /** @nocollapse */ static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.12", ngImport: i0, type: DxiFormTabModule, imports: [DxiFormTabComponent], exports: [DxiFormTabComponent] });
    /** @nocollapse */ static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.12", ngImport: i0, type: DxiFormTabModule, imports: [DxiFormTabComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.12", ngImport: i0, type: DxiFormTabModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        DxiFormTabComponent
                    ],
                    exports: [
                        DxiFormTabComponent
                    ],
                }]
        }] });
