// devextreme-angular/esm2015/ui/nested/base/charts-color.js
/*!
 * devextreme-angular
 * Version: 23.1.13
 * Build date: Mon Oct 28 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
/* tslint:disable:max-line-length */
import { NestedOption } from 'devextreme-angular/core';
import { Component, } from '@angular/core';
import * as i0 from "@angular/core";
export class DxoChartsColor extends NestedOption {
    get base() {
        return this._getOption('base');
    }
    set base(value) {
        this._setOption('base', value);
    }
    get fillId() {
        return this._getOption('fillId');
    }
    set fillId(value) {
        this._setOption('fillId', value);
    }
}
/** @nocollapse */ DxoChartsColor.ɵfac = /** @pureOrBreakMyCode */ function () { let ɵDxoChartsColor_BaseFactory; return function DxoChartsColor_Factory(t) { return (ɵDxoChartsColor_BaseFactory || (ɵDxoChartsColor_BaseFactory = i0.ɵɵgetInheritedFactory(DxoChartsColor)))(t || DxoChartsColor); }; }();
/** @nocollapse */ DxoChartsColor.ɵcmp = /** @pureOrBreakMyCode */ i0.ɵɵdefineComponent({ type: DxoChartsColor, selectors: [["ng-component"]], features: [i0.ɵɵInheritDefinitionFeature], decls: 0, vars: 0, template: function DxoChartsColor_Template(rf, ctx) { }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(DxoChartsColor, [{
        type: Component,
        args: [{
                template: ''
            }]
    }], null, null); })();
