// @thisissoon/angular-inviewport/esm2015/app/window/window-mock.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const WINDOW_MOCK_WITHOUT_INTERSECTION_OBSERVER = {
    closed: null,
    defaultStatus: null,
    document: null,
    frameElement: null,
    frames: null,
    history: null,
    innerHeight: null,
    innerWidth: null,
    length: null,
    localStorage: null,
    location: null,
    name: null,
    navigator: null,
    opener: null,
    outerHeight: null,
    outerWidth: null,
    pageXOffset: null,
    pageYOffset: null,
    parent: null,
    screen: null,
    screenLeft: null,
    screenTop: null,
    screenX: null,
    screenY: null,
    sessionStorage: null,
    scrollX: null,
    scrollY: null,
    self: null,
    status: null,
    top: null,
    addEventListener: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    alert: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    atob: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    blur: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    btoa: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    clearInterval: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    clearTimeout: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    close: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    confirm: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    focus: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    getComputedStyle: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    getSelection: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    matchMedia: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    moveBy: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    moveTo: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    open: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    print: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    prompt: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    resizeBy: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    resizeTo: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    removeEventListener: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    scroll: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    scrollBy: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    scrollTo: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    setInterval: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    setTimeout: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null),
    stop: (/**
     * @param {...?} args
     * @return {?}
     */
    (...args) => null)
};
/** @type {?} */
export const WINDOW_MOCK = Object.assign({}, WINDOW_MOCK_WITHOUT_INTERSECTION_OBSERVER, { IntersectionObserver: (/**
     * @param {...?} args
     * @return {?}
     */
    function (...args) {
        this.observe = (/**
         * @return {?}
         */
        () => null);
        this.unobserve = (/**
         * @return {?}
         */
        () => null);
    }), IntersectionObserverEntry: class IntersectionObserverEntryMock {
    } });
