// ng-flex-layout/fesm2022/ng-flex-layout.mjs
import * as i0 from '@angular/core';
import { Version, PLATFORM_ID, Inject, NgModule } from '@angular/core';
import { DEFAULT_CONFIG, LAYOUT_CONFIG, BREAKPOINT, SERVER_TOKEN } from 'ng-flex-layout/core';
export * from 'ng-flex-layout/core';
import { ExtendedModule } from 'ng-flex-layout/extended';
export * from 'ng-flex-layout/extended';
import { FlexModule } from 'ng-flex-layout/flex';
export * from 'ng-flex-layout/flex';
import { GridModule } from 'ng-flex-layout/grid';
export * from 'ng-flex-layout/grid';
import { isPlatformServer } from '@angular/common';

var name = "ng-flex-layout";
var version = "20.0.0";
var exports = {
	"./mq": {
		sass: "./core/sass/_layout-bp.scss"
	},
	"./_mq": {
		sass: "./core/sass/_layout-bp.scss"
	}
};
var repository = {
	type: "git",
	url: "https://github.com/alessiobianchini/flex-layout.git"
};
var keywords = [
	"angular",
	"flex-layout",
	"flexbox css",
	"media query",
	"breakpoints"
];
var license = "MIT";
var bugs = {
	url: "https://github.com/alessiobianchini/flex-layout/issues"
};
var homepage = "https://github.com/alessiobianchini/flex-layout#readme";
var dependencies = {
	tslib: "^2.8.1"
};
var peerDependencies = {
	"@angular/cdk": "^20.0.0",
	"@angular/common": "^20.0.0",
	"@angular/core": "^20.0.0",
	"@angular/platform-browser": "^20.0.0",
	rxjs: "^6.5.3 || ^7.4.0"
};
var type = "module";
var packageJson = {
	name: name,
	version: version,
	exports: exports,
	repository: repository,
	keywords: keywords,
	license: license,
	bugs: bugs,
	homepage: homepage,
	dependencies: dependencies,
	peerDependencies: peerDependencies,
	type: type
};

/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
/** Current version of Angular Flex-Layout. */
const VERSION = new Version(packageJson.version);

/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
/**
 * FlexLayoutModule -- the main import for all utilities in the Angular Layout library
 * * Will automatically provide Flex, Grid, and Extended modules for use in the application
 * * Can be configured using the static withConfig method, options viewable on the Wiki's
 *   Configuration page
 */
class FlexLayoutModule {
    /**
   * Initialize the FlexLayoutModule with a set of config options,
   * which sets the corresponding tokens accordingly
   */
    static withConfig(configOptions, breakpoints = []) {
        return {
            ngModule: FlexLayoutModule,
            providers: configOptions.serverLoaded ?
                [
                    { provide: LAYOUT_CONFIG, useValue: { ...DEFAULT_CONFIG, ...configOptions } },
                    { provide: BREAKPOINT, useValue: breakpoints, multi: true },
                    { provide: SERVER_TOKEN, useValue: true },
                ] : [
                { provide: LAYOUT_CONFIG, useValue: { ...DEFAULT_CONFIG, ...configOptions } },
                { provide: BREAKPOINT, useValue: breakpoints, multi: true },
            ]
        };
    }
    constructor(serverModuleLoaded, platformId) {
        if (isPlatformServer(platformId) && !serverModuleLoaded) {
            console.warn('Warning: Flex Layout loaded on the server without FlexLayoutServerModule');
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.3", ngImport: i0, type: FlexLayoutModule, deps: [{ token: SERVER_TOKEN }, { token: PLATFORM_ID }], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "20.0.3", ngImport: i0, type: FlexLayoutModule, imports: [FlexModule, ExtendedModule, GridModule], exports: [FlexModule, ExtendedModule, GridModule] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "20.0.3", ngImport: i0, type: FlexLayoutModule, imports: [FlexModule, ExtendedModule, GridModule, FlexModule, ExtendedModule, GridModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.3", ngImport: i0, type: FlexLayoutModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [FlexModule, ExtendedModule, GridModule],
                    exports: [FlexModule, ExtendedModule, GridModule]
                }]
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [SERVER_TOKEN]
                }] }, { type: Object, decorators: [{
                    type: Inject,
                    args: [PLATFORM_ID]
                }] }] });

/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
/**
 * @module
 * @description
 * Entry point for all public APIs of Angular Flex-Layout.
 */

/**
 * Generated bundle index. Do not edit.
 */

export { FlexLayoutModule, VERSION };
