// @thisissoon/angular-inviewport/esm5/app/in-viewport/in-viewport.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostBinding, EventEmitter, Output, Inject, Input } from '@angular/core';
import { WINDOW } from '../window/window-token';
/**
 * A simple lightweight library for Angular with that detects when an
 * element is within the browsers viewport and adds a `in-viewport` or
 * `not-in-viewport` class to the element.
 *
 * \@example
 * ```html
 * <p
 *  class="foo"
 *  snInViewport
 *  (inViewportChange)="myEventHandler($event)">
 *  Amet tempor excepteur occaecat nulla.
 * </p>
 * ```
 */
// @dynamic
var InViewportDirective = /** @class */ (function () {
    function InViewportDirective(el, window) {
        this.el = el;
        this.window = window;
        this.inViewportChange = new EventEmitter();
        this.hasIntersectionObserver = this.intersectionObserverFeatureDetection();
    }
    Object.defineProperty(InViewportDirective.prototype, "isInViewport", {
        get: /**
         * @return {?}
         */
        function () {
            return this.inViewport;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InViewportDirective.prototype, "isNotInViewport", {
        get: /**
         * @return {?}
         */
        function () {
            return !this.inViewport;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    InViewportDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (!this.hasIntersectionObserver) {
            this.inViewport = true;
            this.inViewportChange.emit(this.inViewport);
        }
    };
    /**
     * @return {?}
     */
    InViewportDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this.hasIntersectionObserver) {
            /** @type {?} */
            var IntersectionObserver_1 = this.window['IntersectionObserver'];
            this.observer = new IntersectionObserver_1(this.intersectionObserverCallback.bind(this), this.inViewportOptions);
            this.observer.observe(this.el.nativeElement);
        }
    };
    /**
     * @return {?}
     */
    InViewportDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.observer) {
            this.observer.unobserve(this.el.nativeElement);
        }
    };
    /**
     * @param {?} entries
     * @return {?}
     */
    InViewportDirective.prototype.intersectionObserverCallback = /**
     * @param {?} entries
     * @return {?}
     */
    function (entries) {
        var _this = this;
        entries.forEach((/**
         * @param {?} entry
         * @return {?}
         */
        function (entry) {
            if (_this.inViewport === entry.isIntersecting)
                return;
            _this.inViewport = entry.isIntersecting;
            _this.inViewportChange.emit(_this.inViewport);
        }));
    };
    /**
     * @private
     * @return {?}
     */
    InViewportDirective.prototype.intersectionObserverFeatureDetection = /**
     * @private
     * @return {?}
     */
    function () {
        // Exits early if all IntersectionObserver and IntersectionObserverEntry
        // features are natively supported.
        if ('IntersectionObserver' in this.window &&
            'IntersectionObserverEntry' in this.window) {
            // Minimal polyfill for Edge 15's lack of `isIntersecting`
            // See: https://github.com/w3c/IntersectionObserver/issues/211
            if (!('isIntersecting' in
                this.window['IntersectionObserverEntry']['prototype'])) {
                Object.defineProperty(this.window['IntersectionObserverEntry']['prototype'], 'isIntersecting', {
                    get: (/**
                     * @return {?}
                     */
                    function () {
                        return this.intersectionRatio > 0;
                    })
                });
            }
            return true;
        }
        return false;
    };
    InViewportDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[snInViewport]',
                    exportAs: 'snInViewport'
                },] }
    ];
    /** @nocollapse */
    InViewportDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Window, decorators: [{ type: Inject, args: [WINDOW,] }] }
    ]; };
    InViewportDirective.propDecorators = {
        inViewportOptions: [{ type: Input }],
        inViewportChange: [{ type: Output }],
        isInViewport: [{ type: HostBinding, args: ['class.sn-viewport--in',] }],
        isNotInViewport: [{ type: HostBinding, args: ['class.sn-viewport--out',] }]
    };
    return InViewportDirective;
}());
export { InViewportDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    InViewportDirective.prototype.inViewport;
    /**
     * @type {?}
     * @private
     */
    InViewportDirective.prototype.hasIntersectionObserver;
    /** @type {?} */
    InViewportDirective.prototype.inViewportOptions;
    /** @type {?} */
    InViewportDirective.prototype.inViewportChange;
    /** @type {?} */
    InViewportDirective.prototype.observer;
    /**
     * @type {?}
     * @private
     */
    InViewportDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    InViewportDirective.prototype.window;
}
